Docker 101 (#382)
arcus/education_modules: docker_101/docker_101.md

@end

resource1_name: Docker 101 Tutorial
resource1_description: From the Docker website: "In this self-paced, hands-on tutorial, you will learn how to build images, run containers, use volumes to persist data and mount in source code, and define your application using Docker Compose. You’ll even learn about a few advanced topics, such as networking and image building best practices."
resource1_wellvetted: true
resource1_wellvetted_text:  The Docker 101 tutorial is an [open source project](https://github.com/docker/getting-started), so it has many authors, but it is sponsored and hosted by Docker Inc, the company that produces the docker software. The open source nature of the tutorial also helps ensure that any errors or problems can be caught and addressed quickly.
resource1_maintained: true
resource1_maintained_text: This tutorial is maintained by Docker Inc, so we expect that it will remain up to date whenever changes are implemented in docker.
resource1_stablesupport: true
resource1_stablesupport_text: This is hosted on the Docker website, and it is a popular and widely-shared tutorial. We expect it will continue to be available for the foreseeable future.

> resource1_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

Author: It isn't actually supposed to, for the a11y issues. Originally that was the thought, but we had a meeting and Rose F pointed out that it's pretty problematic for us to assert that anything is fully "accessible" since that's something that depends on the user and context. So the a11y issues thing is just supposed to print on its own beneath the checklist. 
Right now it has some pesky brackets `{}` that are creeping in there from I dont know where, but that's a separate problem...

> resource1_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

Reviewer: Oh okay I see! Is it a thing we're letting slide at the moment in the interest of moving forward with modules, or should that be a task we work on in the near future?

> resource1_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

Author: I think we should fix it ASAP, but either way it's a problem in the macros_wrapper file, not anything in this actual module. So it would be a separate PR from this one.

> resource1_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

Reviewer: Oh okay that makes sense!

resource2_wellvetted_text: This article was written by expert authors and published in PLOS Computational Biology, a peer reviewed scientific journal (you can read about their <a href="https://journals.plos.org/ploscompbiol/s/editorial-and-peer-review-process">editorial and peer review process</a> on their website). 
resource2_maintained: true
resource2_maintained_text: PLOS Computational Biology links any corrections, expressions of concern, or retractions to the affected article's PLOS web page, so any important updates should be available there (for more details read <a href="https://journals.plos.org/ploscompbiol/s/corrections-expressions-of-concern-and-retractions">PLOS's policy on posting corrections</a>).
resource2_stablesupport: true
resource2_stablesupport_text: This article has an assigned DOI (digital identifier of an object), which helps to ensure it will continue to be available (read more <a href="https://www.doi.org/the-identifier/what-is-a-doi/">about DOIs at doi.org</a>). 
resource2_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

@module_structure

> 1. Do the first half of the "Docker 101" tutorial (on docker.com).

Author: I'd say no --- this is kind of a new thing that has come up for wrapper modules in particular, but I have the feeling that we want to really guide learners through **when** to go to the external resource. I definitely don't want someone to click a docker link in the overview and then end up looking at the tutorial right away before we've had the chance to contextualize it. 
I think, though, that it's weird to have something like this that looks like it should be a link but isn't. What do you think about changing it to "(on the Docker website)"?

2. Return to this module to complete the "Docker 101 Quiz" to check your understanding and consolidate your knowledge.
3. Read the article "Ten simple rules for writing Dockerfiles for reproducible data science" by Nüst et al. 2020 (on plos.org).
4. Return to this module for the "Docker for Research Quiz", and for the final sections of the module.
@end

@version_history

@end

import: https://raw.githubusercontent.com/arcus/education_modules/main/_module_templates/macros.md
import: https://raw.githubusercontent.com/arcus/education_modules/main/_module_templates/macros_wrapper.md
-->

# Getting Started with Docker for Research

> @overview

Author: I think I fixed this --- lia was tripping on the empty `version_history` macro

resource2_wellvetted_text: This article was written by expert authors and published in PLOS Computational Biology, a peer reviewed scientific journal (you can read about their <a href="https://journals.plos.org/ploscompbiol/s/editorial-and-peer-review-process">editorial and peer review process</a> on their website). 
resource2_maintained: true
resource2_maintained_text: PLOS Computational Biology links any corrections, expressions of concern, or retractions to the affected article's PLOS web page, so any important updates should be available there (for more details read <a href="https://journals.plos.org/ploscompbiol/s/corrections-expressions-of-concern-and-retractions">PLOS's policy on posting corrections</a>).
resource2_stablesupport: true
resource2_stablesupport_text: This article has an assigned DOI (digital identifier of an object), which helps to ensure it will continue to be available (read more <a href="https://www.doi.org/the-identifier/what-is-a-doi/">about DOIs at doi.org</a>). 
resource2_a11y_issues: No known issues with accessibility, but we may have missed something. If you encounter an issue, please [let us know](#feedback)!

@module_structure

> 1. Do the first half of the "Docker 101" tutorial (on docker.com).

Author: ```suggestion
1. Do the first half of the "Docker 101" tutorial (on the Docker website).
```

<!--

author:   Rose Hartman
email:    hartmanr1@chop.edu
version:  1.0.0
current_version_description: Initial version
module_type: wrapper
docs_version: 1.0.0
language: en
narrator: UK English Female
mode: Textbook

title: Getting Started with Docker for Research

comment: This tutorial combines a hands-on interactive Docker tutorial published by Docker Inc with an academic article outlining best practices for using Docker for research. 

> long_description: If you've been curious about how to use Docker for your research, this module is a great place to start. The Docker 101 tutorial is a popular, hands-on approach to learning Docker that will get you using containers right away, so you can learn by doing. To help you bridge the gap between basic Docker use and best practices for using Docker in research, we also link to an article outlining 10 rules to help you create great containers for research, and lists of ready-to-use Docker images for a variety of analysis workflows. This module includes running and editing comamnds in the terminal, so you'll need some basic familiarity with bash, but it is otherwise appropriate for beginniners. No prior experience with Docker or containers is assumed. 

Reviewer: ```suggestion
long_description: If you've been curious about how to use Docker for your research, this module is a great place to start. The Docker 101 tutorial is a popular, hands-on approach to learning Docker that will get you using containers right away, so you can learn by doing. To help you bridge the gap between basic Docker use and best practices for using Docker in research, we also link to an article outlining 10 rules to help you create great containers for research, and lists of ready-to-use Docker images for a variety of analysis workflows. This module includes running and editing commands in the terminal, so you'll need some basic familiarity with bash, but it is otherwise appropriate for beginners. No prior experience with Docker or containers is assumed. 
```

Or a container that takes a set of RMarkdown files and images and compiles them into a pdf version of the article you're drafting. 
Or a container that trains a machine learning model on a provided dataset. 

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

As you consider the tutorial, imagine the kinds of coding projects you typically work on --- whatever those are --- rather than the example code they provide. 

</div>

### Before you Start

To do the Docker 101 tutorial, you'll need a couple things set up: 

- Create an account (it's free) at Docker Hub ([Docker Hub signup page](https://hub.docker.com/signup/)).
- Install [Docker Desktop](https://docs.docker.com/desktop/) on your computer ([installation instructions for Windows](https://docs.docker.com/desktop/install/windows-install/), [installation instructions for Mac](https://docs.docker.com/desktop/install/mac-install/), and [installation instructions for Linux](https://docs.docker.com/desktop/install/linux-install/)).

<div class = "help">
<b style="color: rgb(var(--color-highlight));">Troubleshooting help</b><br>

> To check that you have successfully installed Docker on your computer, open the comand line and run the following: 

Reviewer: ```suggestion
To check that you have successfully installed Docker on your computer, open the command line and run the following: 
```

The command `docker run` starts a container that you've already built. 

</div>
***

Once you've built a Docker image, what is the easiest way to share it?

[( )] Send the Dockerfile via email
[(X)] Share the image via an online repository
[( )] Either of the above will work
***
<div class = "answer">

The best way to share a container image is with the [Docker Hub](https://hub.docker.com/) repository. For a review, see the "Sharing our App" section of the tutorial.

Note that sending just the Dockerfile to someone isn't enough for them to be able to build the container image (just as sending a recipe to someone won't let them cook the dish unless you also send the ingredients), so that won't work. 

</div>
***

## Using Docker in Research

Most of the tutorials and explanations you'll find online about Docker (including the Docker 101 tutorial you just worked on) focus on its use in software development. 
Importantly, most of the example Dockerfiles you'll find online are also written by and for developers. 
As someone new to Docker (and even for experienced users!), being able to copy existing Dockerfiles is an important time saver.

> Although you might be able to get these images to work for your research needs, there are a lot of special considerations that come up when you're using containers for reproducible research rather than software develpoment. 

Reviewer: ```suggestion
Although you might be able to get these images to work for your research needs, there are a lot of special considerations that come up when you're using containers for reproducible research rather than software development. 
```

<b style="color: rgb(var(--color-highlight));">External Content</b><br>

Next, read [Ten simple rules for writing Dockerfiles for reproducible data science](https://doi.org/10.1371/journal.pcbi.1008316) by Nüst et al. (2020). 

<br>
<b style="color: rgb(var(--color-highlight));">Read the whole article, then return here to finish this module.</b>
</div>

### Quiz: Docker for Research

True or False: The authors recommend including your dataset in the container, to make your analysis fully reproducible.

[( )] True
[(X)] False
***
<div class = "answer">

They recommend storing datafiles outside of the container and using bind mounts to attach the data just when you run the container, rather than building it into the image. 
See [Rule 7: Mount datasets at run time](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008316#sec020). 

For more guidance on using bind mounts, see the [Docker documentation on bind mounts](https://docs.docker.com/storage/bind-mounts/).

</div>
***

Which of the following would the authors say is more important?

[(X)] Making your Dockerfile as explicit as possible, so the layers in the image are easier to see

> [( )] Making your Dockerfile as efficient as posisble, so the image is smaller and more portable

Reviewer: ```suggestion
[( )] Making your Dockerfile as efficient as possible, so the image is smaller and more portable
```

***
<div class = "answer">

See [Rule 3: Format for clarity](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008316#sec009). 
In particular:

> In general, if your container is mostly software dependencies, you should not need to worry about image size because (a) your data is likely to have much larger storage requirements; and (b) transparency and inspectability outweigh storage concerns in data science. 

</div>
***

What character is used to start a comment line in a Dockerfile?

[[#]]
<script>
  let input = "@input".trim().toLowerCase();
  input == "#";
</script>
***
<div class = "answer">

See [Rule 4: Document within the Dockerfile](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008316#sec010) for examples.

</div>
***

> True or False: For a container with complex anlaysis code, the authors recommend writing your Dockerfile so that the whole analysis workflow runs automatically when you start the container.

Reviewer: ```suggestion
True or False: For a container with complex analysis code, the authors recommend writing your Dockerfile so that the whole analysis workflow runs automatically when you start the container.
```
